---
title: "Figure 2"
author: "Matt McCoy"
date: "January 11, 2020"
output: html_document
---
  
Load libraries.
```{r}
library(biomaRt)
library(ggplot2)
library(reshape2)
library(intervals)
library(gridExtra)
library(taxonomizr)
library(ape)
library(phylobase)
library(phytools)
library(Hmisc)
library(RColorBrewer)
```

Load functions.
```{r}
source("../code/get_attributes.R")
source("../code/get_gene_table.R")
source("../code/get_ensembl.R")
source("../code/get_species_table.R")
```

Load ensembl databases.
```{r}
chordata_ensembl <- get_ensembl("chordata")
metazoa_ensembl <- get_ensembl("metazoa")
plants_ensembl <- get_ensembl("plants")
protists_ensembl <- get_ensembl("protists")
fungi_ensembl <- get_ensembl("fungi")
```

Calculate protein-coding feature estimates in Plantae.
```{r}
# To obtain Taxon IDs, download species table from https://plants.ensembl.org/species.html.
filename <- "./data/plants_species_table.csv"
plants.species.conversion <- read.table(file = filename, header = TRUE, sep = ",")
plants.datasets <- listDatasets(plants_ensembl) # list all available species
filename2 <- "./data/plants.datasets.csv"
if(!file.exists(filename2)){
  plants.datasets$description <- gsub("\\ genes.*","", plants.datasets$description) # reformat names
  plants.datasets$Taxon.ID <- plants.species.conversion$Taxon.ID[match(as.character(plants.datasets$description), as.character(plants.species.conversion$Name))] # match Taxon IDs
  write.csv(plants.datasets, file = filename2)
} else{
  plants.datasets <- read.csv(file = filename2, sep = ",", header = TRUE, row.names = 1, stringsAsFactors = FALSE)
}

plants.species.table <- get_species_table(plants.datasets)
```

Now in Protista.
```{r}
# To obtain Taxon IDs, download species table from https://protists.ensembl.org/species.html.
filename <- "./data/protists_species_table.csv"
protists.species.conversion <- read.table(file = filename, header = TRUE, sep = ",")
protists.datasets <- listDatasets(protists_ensembl) # list all available species
filename2 <- "./data/protists.datasets.csv"
if(!file.exists(filename2)){
  protists.datasets$description <- gsub("\\ genes.*","", protists.datasets$description) # reformat names
  protists.datasets$Taxon.ID <- protists.species.conversion$Taxon.ID[match(as.character(protists.datasets$description), as.character(protists.species.conversion$Name))] # match Taxon IDs
  write.csv(protists.datasets, file = filename2)
} else{
  protists.datasets <- read.csv(file = filename2, sep = ",", header = TRUE, row.names = 1, stringsAsFactors = FALSE)
}

protists.species.table <- get_species_table(protists.datasets)
```

Now in Fungi.
```{r}
# To obtain Taxon IDs, download species table from http://fungi.ensembl.org/species.html.
filename <- "./data/fungi_species_table.csv"
fungi.species.conversion <- read.table(file = filename, header = TRUE, sep = ",")
fungi.datasets <- listDatasets(fungi_ensembl) # list all available species
filename2 <- "./data/fungi.datasets.csv"
if(!file.exists(filename2)){
  fungi.datasets$description <- gsub("\\ genes.*","", fungi.datasets$description) # reformat names
  fungi.datasets$Taxon.ID <- fungi.species.conversion$Taxon.ID[match(as.character(fungi.datasets$description), as.character(fungi.species.conversion$Name))] # match Taxon IDs
  write.csv(fungi.datasets, file = filename2)
} else{
  fungi.datasets <- read.csv(file = filename2, sep = ",", header = TRUE, row.names = 1, stringsAsFactors = FALSE)
}

fungi.species.table <- get_species_table(fungi.datasets)
```

Now in Metazoa.
```{r}
# To obtain Taxon IDs, download species table from http://metazoa.ensembl.org/species.html.
filename1 <- "./data/metazoa_species_table.csv"
metazoa.species.conversion <- read.table(file = filename1, header = TRUE, sep = ",")
metazoa.datasets <- listDatasets(metazoa_ensembl) # list all available species
filename2 <- "./data/metazoa.datasets.csv"
if(!file.exists(filename2)){
  metazoa.datasets$description <- gsub("\\ genes.*","", metazoa.datasets$description) # reformat names
  metazoa.datasets$Taxon.ID <- metazoa.species.conversion$Taxon.ID[match(as.character(metazoa.datasets$description), as.character(metazoa.species.conversion$Name))] # match Taxon IDs
  write.csv(metazoa.datasets, file = filename2)
} else{
  metazoa.datasets <- read.csv(file = filename2, sep = ",", header = TRUE, row.names = 1, stringsAsFactors = FALSE)
}

metazoa.species.table <- get_species_table(metazoa.datasets)
```

Now in Chordata.
```{r}
# To obtain Taxon IDs, download species table from http://uswest.ensembl.org/info/about/species.html. I added some missing taxonomy information.
filename1 <- "./data/chordata_species_table.csv"
chordata.species.conversion <- read.csv(file = filename1, header = TRUE)
chordata.datasets <- listDatasets(chordata_ensembl) # list all available species
filename2 <- "./data/chordata.datasets.csv"
if(!file.exists(filename2)){
  chordata.datasets$description <- gsub("\\ genes.*","", chordata.datasets$description) # reformat names
  # Remove non-chordates
  chordata.datasets <- chordata.datasets[! chordata.datasets$dataset %in% c("celegans_gene_ensembl", "dmelanogaster_gene_ensembl", "scerevisiae_gene_ensembl"),]
  # Remove duplicate hamster genome and male mole rat genome
  chordata.datasets <- chordata.datasets[! chordata.datasets$dataset %in% c("cchok1gshd_gene_ensembl", "hmale_gene_ensembl"),]
  # Match species names
  chordata.datasets$species <- chordata.species.conversion$Scientific.name[match(as.character(chordata.datasets$description), as.character(chordata.species.conversion$Common.name))]
  # Match remaining missing species names.
  chordata.datasets$species[is.na(chordata.datasets$species)] <- chordata.species.conversion$Scientific.name[match(subset(chordata.datasets, is.na(species))$version, as.character(chordata.species.conversion$Ensembl.Assembly))]
  chordata.datasets
  # Match Taxon IDs
  chordata.datasets$Taxon.ID <- chordata.species.conversion$Taxon.ID[match(as.character(chordata.datasets$species), as.character(chordata.species.conversion$Scientific.name))]
  chordata.datasets[,-5] <- as.character(chordata.datasets[,-5])
  write.csv(chordata.datasets, file = filename2)
} else{
  chordata.datasets <- read.csv(file = filename2, sep = ",", header = TRUE, row.names = 1, stringsAsFactors = FALSE)
}

chordata.species.table <- get_species_table(chordata.datasets)
```

```{r}
# Combine and compare across different clades.
species.table <- rbind(plants.species.table, metazoa.species.table, chordata.species.table, fungi.species.table, protists.species.table)

# Append full Taxonomy
prepareDatabase('accessionTaxa.sql')
taxa <- getTaxonomy(species.table$Taxon.ID, 'accessionTaxa.sql')
species.table <- merge(species.table, taxa, by = "species")
# Fill in missing chordate information
species.table$class <- as.character(species.table$class)
species.table$class[species.table$species %in% subset(species.table, is.na(class) & clade == "Chordata")$species] <- c("Reptilia", "Myxini", "Reptilia", "Sarcopterygii", "Reptilia", "Hyperoartia", "Reptilia")
```

Figure 2A. species.table is used for the median long gene length. Timetree was used to generate newick tree. iTOL was used to generate plot.

```{r}
# Mean and standard deviations reported within main text
tapply(species.table$median_gene_length, species.table$clade, mean)
tapply(species.table$median_gene_length, species.table$clade, sd)
```

###########
# All animals.
###########

All genes.
```{r}
mean(subset(species.table, clade == "Chordata"|clade == "Metazoa")$median_gene_length)
sd(subset(species.table, clade == "Chordata"|clade == "Metazoa")$median_gene_length)
```


```{r}
# Top 10% longest genes in each species
mean(subset(species.table, clade == "Chordata"|clade == "Metazoa")$median_long_gene_length)
sd(subset(species.table, clade == "Chordata"|clade == "Metazoa")$median_long_gene_length)
```

```{r}
# Top 10% longest genes in chordates
mean(subset(species.table, clade == "Chordata")$median_long_gene_length)
sd(subset(species.table, clade == "Chordata")$median_long_gene_length)
```

```{r}
# Top 10% longest genes in invertebrates
mean(subset(species.table, clade == "Metazoa")$median_long_gene_length)
sd(subset(species.table, clade == "Metazoa")$median_long_gene_length)
```

```{r}
# All eukaryotes except animals
mean(subset(species.table, clade != "Chordata"&clade != "Metazoa")$median_long_gene_length)
sd(subset(species.table, clade != "Chordata"&clade != "Metazoa")$median_long_gene_length)
```


```{r}
# Ciona long gene lengths
mean(subset(species.table, species == "Ciona savignyi"|species == "Ciona intestinalis")$median_long_gene_length)
sd(subset(species.table, species == "Ciona savignyi"|species == "Ciona intestinalis")$median_long_gene_length)
```


```{r}
# Wilcox signed rank test with continuity corrections for differences in gene lengths
wilcox.test(x = subset(species.table, clade == "Metazoa"|clade == "Chordata")$median_long_gene_length, mu = mean(subset(species.table, species == "Amphimedon queenslandica"|species == "Trichoplax adhaerens")$median_long_gene_length))
wilcox.test(x = subset(species.table, clade == "Metazoa")$median_long_gene_length, mu = mean(subset(species.table, species == "Amphimedon queenslandica"|species == "Trichoplax adhaerens")$median_long_gene_length))
```


```{r}
# Fish gene lengths
mean(subset(species.table, class == "Actinopteri"|order == "Coelacanthiformes")$median_long_gene_length)
sd(subset(species.table, class == "Actinopteri"|order == "Coelacanthiformes")$median_long_gene_length)
```


```{r}
# Bird gene lengths
mean(subset(species.table, class == "Aves")$median_long_gene_length)
sd(subset(species.table, class == "Aves")$median_long_gene_length)
```


```{r}
# Reptile gene lengths
mean(subset(species.table, class == "Reptilia")$median_long_gene_length)
sd(subset(species.table, class == "Reptilia")$median_long_gene_length)
```


```{r}
# Mammals (not Primates) gene lengths
mean(subset(species.table, class == "Mammalia"& order != "Primates")$median_long_gene_length)
sd(subset(species.table, class == "Mammalia"& order != "Primates")$median_long_gene_length)
```


```{r}
# Humans
subset(species.table, species == "Homo sapiens")$median_long_gene_length
```


```{r}
# Primates gene lengths
mean(subset(species.table, order == "Primates" & species != "Homo sapiens")$median_long_gene_length)
sd(subset(species.table, order == "Primates" & species != "Homo sapiens")$median_long_gene_length)
```

Save data.
```{r}
# re-order clade factor levels
species.table$clade = factor(species.table$clade,levels(species.table$clade)[c(4,5,1,2,3)])

# save data
write.csv(species.table, file = "./data/species.table.csv")
```

Plots for Figure 2B-D.
```{r}
g1 <- ggplot(species.table, aes(x = clade, y = median_long_gene_length/1000, fill = clade)) + geom_boxplot() + scale_y_log10() + theme_classic() + annotation_logticks(sides = "l") + ylab("median long gene length (kb)") + xlab("") + theme(legend.position="none", axis.text.x=element_text(angle=45, hjust=1))
region_melt <- melt(species.table[,c(1,3,5,7,8,9,10)], id.vars = c("species", "median_long_gene_length", "clade", "number_of_long_gene_exons", "number_of_exons"))
g2 <- ggplot(region_melt, aes(x = clade, y = value/1000, fill = variable)) + geom_boxplot() + scale_y_log10() + theme_classic() + annotation_logticks(sides = "l") + xlab("") + ylab("Feature length (kb)") + theme(axis.text.x=element_text(angle=45, hjust=1))
g3 <- ggplot(region_melt, aes(x = clade, y = number_of_exons, fill = clade)) + geom_boxplot() + theme_classic() + xlab("") + ylab("Number of exons") + theme(legend.position="none", axis.text.x=element_text(angle=45, hjust=1))
g <- arrangeGrob(g1,g2,g3,ncol = 3)
ggsave(g, filename = "./plots/Figure2B-D.pdf", device = "pdf", width = 30, units = "mm", height = 225, scale = 2, useDingbats = FALSE)
```
